examples/service_worker/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+//! 1. Cache resources
+//! 2. Register the service worker
+//! 3. If the service worker is not yet activated, an even listener will be registered, waiting for the
+//!    state to reach "activated".
+//! 4. When the state reaches "activated", the Notification object will request permission for notifications
+//! 5. If permission is granted, the `PushManager` will subscribe using an example vapid key
+//! 6. Finally, a `PushSubscription` will be returned, containing the information that can be passed to a
+//!    notification back-end server.
+
+#![allow(clippy::cast_possible_truncation, clippy::needless_pass_by_value)]
+
+pub mod errors;
+
+use crate::errors::ServiceWorkerError;
+use seed::{prelude::*, *};
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+struct PushSubscriptionKeys {
+    p256dh: String,
+    auth: String,
+}
+
+#[serde(rename_all = "camelCase")]
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+struct PushSubscription {
+    endpoint: String,
+    expiration_time: Option<String>,
+    keys: PushSubscriptionKeys,
+}
+
+// Register the push manager given a ServiceWorkerRegistration object
+async fn register_push_manager(
+    sw_reg: web_sys::ServiceWorkerRegistration,
+) -> Result<Msg, ServiceWorkerError> {
+    let permission = web_sys::Notification::request_permission()
+        .map_err(ServiceWorkerError::RequestPermission)?;
+    let permission = wasm_bindgen_futures::JsFuture::from(permission)
+        .await
+        .map_err(ServiceWorkerError::RequestPermission)?;
+    let permission: String = JsValue::into_serde(&permission)?;
+
+    if !permission.eq("granted") {
+        return Err(ServiceWorkerError::InvalidPermissions);
+    }
+
+    let manager: web_sys::PushManager = sw_reg
+        .push_manager()
+        .map_err(ServiceWorkerError::RetrievePushManager)?;
+
+    // Using `web-push generate-vapid-keys` the following is generated for this example:
+    // =======================================
+
+    // Public Key:
+    // BPUHCMCC6_WLIQh-eo0Bmh-w0fG5txRVLfjVfOXRcGVfIcQeaMSPAin0Q-WHgxNENK_2NCJykknLX7fKN9XY-QQ
+
+    // Private Key:
+    // F2iuoMyqIQKxCuinBuZKodP-6wnUcrW6tsHfbsxwSUA
+
+    // =======================================
+    let key =
+        "BPUHCMCC6_WLIQh-eo0Bmh-w0fG5txRVLfjVfOXRcGVfIcQeaMSPAin0Q-WHgxNENK_2NCJykknLX7fKN9XY-QQ";
+
+    // In order to subscribe to PushNotifications we need to specify two things:
+    // 1. The application server key
+    // 2. userVisibleOnly MUST be set to true (this used to only apply to chrome but it appears firefox requires it as well).
+    // As of Aug 29, 2020, wasm_bindgen does not currently provide the `userVisibleOnly` property. A PR was submitted that should
+    // make its way into the 0.2.68 release: https://github.com/rustwasm/wasm-bindgen/commit/49dc58e58f0a8b5921eb7602ab72e82ec51e65e4
+    let subscription: JsValue;
+    unsafe {
+        subscription = subscribe(manager.clone(), key).await;
+    }
```

Comment:
We don't need `unsafe`. IDE (at least Rust-analyzer) is lying.
```suggestion
    subscription = subscribe(manager.clone(), key).await;
```

Reply:
Nice, I didn't realize that.